enterprise/backend/src/metabase_enterprise/caching/api.clj

```diff
@@ -0,0 +1,118 @@
+(ns metabase-enterprise.caching.api
+  (:require
+   [compojure.core :refer [GET]]
+   [metabase.api.common :as api]
+   [metabase.api.common.validation :as validation]
+   [metabase.api.routes.common :refer [+auth]]
+   [metabase.events :as events]
+   [metabase.util :as u]
+   [metabase.util.cron :as u.cron]
+   [metabase.util.i18n :refer [tru]]
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2]))
+
+(defn- assert-valid-models [model ids]
+  (if (= model "root")
+    (when-not (zero? (first ids))
+      (throw (ex-info (tru "Root configuration is only valid with model_id = 0") {:status-code 400
+                                                                                  :model_id    (first ids)})))
+    (api/check-404 (t2/select-one (case model
+                                    "database"  :model/Database
+                                    "dashboard" :model/Dashboard
+                                    "question"  :model/Card)
+                                  :id [:in ids]))))
+
+(defn- audit-caching-change! [id target prev new]
+  (events/publish-event!
+   :event/caching-update
+   {:user-id api/*current-user-id*
+    :model   :model/CacheConfig
+    :details {:id             id
+              :model          (:model target)
+              :model_id       (:model_id target)
+              :previous-value prev
+              :next-value     new}}))
+
+(api/defendpoint GET "/"
+  "Return cache configuration."
+  [:as {{:strs [model collection]
+         :or   {model "root"}}
+        :query-params}]
+  {model      (ms/QueryVectorOf [:enum "root" "database" "dashboard" "question"])
+   collection [:maybe ms/PositiveInt]}
+  (validation/check-has-application-permission :setting)
+  (let [items (t2/select :model/CacheConfig
+                         :model [:in model]
+                         {:left-join [:report_card      [:and
+                                                         [:= :model [:inline "question"]]
+                                                         [:= :model_id :report_card.id]
+                                                         [:= :report_card.collection_id collection]]
```

Comment:
We want `report_card.collection_id IS NULL` in SQL, it's cards/dashboards with no collection then.

Reply:
Ah, I didn't realize those where the semantics. I'm a little nervous of having such a specific meaning for nil, which might just mean that the caller omitted the key. What do you think about taking a sentinel value in the API, e.g. "none"?

Reply:
Hmm, that's an interesting idea, I practically proxied DB to frontend. I wonder if we have any prior art so that it won't be too different.

Right now this won't be used - Raffi is working on interface which only gives you control over root and databases, so we'll be able to improve situation later on.

Reply:
OK cool, let's build the bridge once the river starts flowing. I would feel a little more secure about people tricking over this potential surprise if you added something along the lines of the following comment above the `[:maybe ms/PositiveInt]` type for `collection`:

> Note that a nil value means the set of all configurations that are *not* scoped to a collection, rather than the absence of any collection filter.